src/main/scala/scalismo/faces/sampling/face/proposals/SphericalHarmonicsLightProposals.scala

```diff
@@ -264,4 +269,133 @@ object SphericalHarmonicsLightProposals {
+        val c = curSample(x,y)
+        if (c.a > 0  ) {
+          counter = counter + 1
+          val eval = pixelEvaluator.logValue(c.toRGB, target(x, y).toRGB)
+          eval
+        }
+        else
+          Math.log(threshold) - normalizer + 1 // just to be above threshold
+      }, AccessMode.Strict[Double])
+      val thresholded = differenceImage.map(d => if (Math.exp(d - normalizer) > threshold) 1 else 0)
+      val count: Double = thresholded.foldLeft(0.0)((a, b) => a + b) / counter
+      (count, thresholded)
+    }
+
+    // samples according to mask
+    private def maskedSampler(rps: RenderParameter, label: PixelImage[Int]) = {
+      val maskProp: MeshSurfaceProperty[Option[Double]] = labelAsSurfaceProperty(rps, label.withAccessMode(AccessMode.Strict[Int]))
+      MeshSurfaceSampling.sampleAccordingToMask(maskProp.map(v => v.getOrElse(0.0)), nSamplesIllumination) _
+    }
+
+    // puts the label from the image to the mesh as surface property
+    private def labelAsSurfaceProperty(rps: RenderParameter, label: PixelImage[Int]): MeshSurfaceProperty[Option[Double]] = {
+      TextureExtraction.imageAsSurfaceProperty(modelRenderer.instance(rps).shape, rps.pointShader, label.map(i => i.toDouble))
+    }
+
+
+    override def propose(current: (RenderParameter, PixelImage[Int])): (RenderParameter, PixelImage[Int]) = {
+
+      // get renderParameters from tuple, the labels from PixelImage[Int] are not used
+      val rps = current._1
+
+      val mesh: VertexColorMesh3D = modelRenderer.instance(rps)
+      val faceMask = modelRenderer.renderImage(rps)
+      val faceMaskLabel: PixelImage[Int] = faceMask.map(p => if (p.a > 0.01) {
+        1
+      } else {
+        0
+      })
+
+      // RANSAC is an iterative method, those variables are keeping track of the current best estimate
+      // the naming of the parameters is motivated by https://en.wikipedia.org/wiki/Random_sample_consensus
+      var soFarBestModelQuality = 0.0
+      var soFarBestLight: RenderParameter = rps
+      var label = faceMaskLabel
+
+      // while iterations < k {
+      for (counter <- 1 until iterations) {
+        // the sampler to sample points on the surface
+        val points = MeshSurfaceSampling.sampleUniformlyOnSurface(nSamples)(mesh.shape)
+
+        // maybemodel = model parameters fitted to maybeinliers
+        val newLight = shOpt.optimize(rps, points)
+        val newParam: RenderParameter = rps.copy(environmentMap = newLight)
+        val (count, thresholded) = measureModelQuality(newParam)
+
+        // do a full model estimation if the estimation based on few samples is good enough
+        if (count > percentage) {
+          // the full model estimation takes 1000 instead of 30 samples (with standard parameters) but is restricted to estimated mask
+          val allPoints = maskedSampler(rps, thresholded)(mesh.shape)
+          val betterModel = shOpt.optimize(rps, allPoints)
```

Comment:
The lines 370-372 do exactly the same as 362-364. I would suggest to either name the variables consitently (light/parameter) or to directly make a function.